docs/4-functional-programming/homework/src/main.ts

```diff
@@ -1,26 +1,75 @@
-import { Either, fromPromise, ap, right, getOrElse, flatten } from './fp/either';
+import { Either, fromPromise, ap, right, getOrElse, flatten, left } from './fp/either';
 import { pipe } from './fp/utils';
 import { fetchClient, fetchExecutor } from './fetching';
 import { ClientUser, ExecutorUser } from './types';
+import { fromNullable, isSome } from './fp/maybe';
+import { distance } from './utils';
+import { fromCompare, ordNumber } from './fp/ord';
+import { sort } from './fp/array';
 
-type Response<R> = Promise<Either<string, R>>
+type Response<R> = Promise<Either<string, R>>;
 
 const getExecutor = (): Response<ExecutorUser> => fromPromise(fetchExecutor());
-const getClients = (): Response<Array<ClientUser>> => fromPromise(fetchClient());
+const getClients = (): Response<Array<ClientUser>> =>
+  fromPromise(
+    fetchClient().then(clientsArray =>
+      clientsArray.map(client => ({ ...client, demands: fromNullable(client.demands) })),
+    ),
+  );
 
 export enum SortBy {
   distance = 'distance',
   reward = 'reward',
 }
 
-export const show = (sortBy: SortBy) => (clients: Array<ClientUser>) => (executor: ExecutorUser): Either<string, string> => {
+export const show =
+  (sortBy: SortBy) =>
+  (clients: Array<ClientUser>) =>
+  (executor: ExecutorUser): Either<string, string> => {
+    const reachableClients = clients.filter(client => {
+      return isSome(client.demands)
+        ? client.demands.value.some(demand => executor.possibilities.includes(demand))
+        : client;
+    });
 
-};
+    const callableClientsSort =
+      sortBy === SortBy.distance
+        ? fromCompare((a: ClientUser, b: ClientUser) => {
+            return ordNumber.compare(
+              distance(a.position, executor.position),
+              distance(b.position, executor.position),
+            );
+          })
+        : fromCompare((a: ClientUser, b: ClientUser) => ordNumber.compare(b.reward, a.reward));
 
-export const main = (sortBy: SortBy): Promise<string> => (
-  Promise
-    .all([getClients(), getExecutor()]) // Fetch clients and executor
-    .then(([clients, executor]) => (
+    const getReachableClientsTable = (reachableClients: Array<ClientUser>) => {
+      return reachableClients.reduce((accumulator: string, reachableClient: ClientUser) => {
+        return (
+          accumulator +
+          `\nname: ${reachableClient.name}, distance: ${distance(
+            executor.position,
+            reachableClient.position,
+          )}, reward: ${reachableClient.reward}`
+        );
+      }, '');
+    };
+
+    return !reachableClients.length
+      ? left('This executor cannot meet the demands of any client!')
+      : reachableClients.length === clients.length
+      ? right('This executor meets all demands of all clients!')
```

Comment:
I believe the sorted clients should also be printed if executor meets demands of all of them